mlir/lib/Dialect/MIOpenOps/CppOutput/gridwise_convolution_implicit_gemm_v4r4_gen_xdlops.cpp

```diff
@@ -30,444 +30,204 @@ namespace {
-      } else {
-        srcDataPerRead_Gemm = 1;
-      }
-  
-      // calculate threadwise copy size
-      const int64_t b_data_per_thread_copy = (param[gemmKPerBlock] * param[gemmNPerBlock]) / blockSize;
-  
-      if(!(b_data_per_thread_copy > 0))
-          return std::make_tuple(-1, -1, -1, -1, false);
-  
-      // GemmBBlockCopySrcDataPerRead_GemmN also bounded by size of threadwise copy
-      srcDataPerRead_Gemm = gcd(srcDataPerRead_Gemm, b_data_per_thread_copy);
-
-      const int64_t b_data_per_thread_copy_gemm_vectorized = srcDataPerRead_Gemm;
-      const int64_t b_data_per_thread_copy_gemm_nonvectorized =
-          b_data_per_thread_copy / b_data_per_thread_copy_gemm_vectorized;
-  
-      int64_t b_data_per_thread_copy_gemmk = 0;
-      int64_t b_data_per_thread_copy_gemmn = 0;
-      if (gemmNVectorizable) {
-        b_data_per_thread_copy_gemmk = b_data_per_thread_copy_gemm_nonvectorized;
-        b_data_per_thread_copy_gemmn = b_data_per_thread_copy_gemm_vectorized;
-      } else {
-        b_data_per_thread_copy_gemmk = b_data_per_thread_copy_gemm_vectorized;
-        b_data_per_thread_copy_gemmn = b_data_per_thread_copy_gemm_nonvectorized;
-      }
+  LogicalResult calculateGemmABlockCopyPerformanceParameters(
+      InitParamsXDL *param, ConvolutionContext &ctx, DerivedParams &derived) {
+    int64_t blockSize = obtainBlockSize(*param, waveSize);
+    return calculateInputDerivedParams(param, blockSize, ctx, true, derived);
+  }
 
-      // GemmBBlockCopyDstDataPerWrite_GemmN also bounded by size of threadwise copy
-      dstDataPerWrite_GemmN = gcd(dstDataPerWrite_GemmN, b_data_per_thread_copy_gemmn);
-  
-      // calculate blockwise copy thread cluster lengths
-      clusterLengths_GemmK = param[gemmKPerBlock] / b_data_per_thread_copy_gemmk;
-      clusterLengths_GemmN = param[gemmNPerBlock] / b_data_per_thread_copy_gemmn;
-  
-      if(!(clusterLengths_GemmK > 0 && clusterLengths_GemmN > 0))
-          return std::make_tuple(-1, -1, -1, -1, false);
-  
-      //llvm::errs() << "======================\n";
-      //llvm::errs() << "Matrix B\n";
-      //llvm::errs() << "gemmK Vectorizable: " << gemmKVectorizable << "\n";
-      //llvm::errs() << "gemmN Vectorizable: " << gemmNVectorizable << "\n";
-      //llvm::errs() << "cluster lengths gemmK: " << clusterLengths_GemmK << "\n";
-      //llvm::errs() << "cluster lengths gemmN: " << clusterLengths_GemmN << "\n";
-      //llvm::errs() << "data per read: " << srcDataPerRead_Gemm << "\n";
-      //llvm::errs() << "data per write: " << dstDataPerWrite_GemmN << "\n";
-      //llvm::errs() << "======================\n";
-
-      return std::make_tuple(clusterLengths_GemmK,
-                             clusterLengths_GemmN,
-                             srcDataPerRead_Gemm,
-                             dstDataPerWrite_GemmN,
-                             true);
+  // TBD: review logic here as they may be tied to NCHW layout.
```

Comment:
is this comment still true?

Reply:
Will remove all TBDs because all the functions now handles more than one layout. I will wait for @zjing14 to either review it or revise it should there be any issues in those vectorization logic.